uefi/firmwarefile.go

```diff
@@ -0,0 +1,112 @@
+package uefi
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+
+	uuid "github.com/linuxboot/fiano/uuid"
+)
+
+// FVFileType represents the different types possible in an EFI file.
+type FVFileType uint8
+
+// UEFI FV File types.
+const (
+	fvFileTypeAll FVFileType = iota
+	fvFileTypeRaw
+	fvFileTypeFreeForm
+	fvFileTypeSECCore
+	fvFileTypePEICore
+	fvFileTypeDXECore
+	fvFileTypePEIM
+	fvFileTypeDriver
+	fvFileTypeCombinedPEIMDriver
+	fvFileTypeApplication
+	fvFileTypeSMM
+	fvFileTypeVolumeImage
+	fvFileTypeCombinedSMMDXE
+	fvFileTypeSMMCore
+	fvFileTypeSMMStandalone
+	fvFileTypeSMMCoreStandalone
+	fvFileTypeOEMMin   FVFileType = 0xC0
+	fvFileTypeOEMMax   FVFileType = 0xDF
+	fvFileTypeDebugMin FVFileType = 0xE0
+	fvFileTypeDebugMax FVFileType = 0xEF
+	fvFileTypePad      FVFileType = 0xF0
+	fvFileTypeFFSMin   FVFileType = 0xF0
+	fvFileTypeFFSMax   FVFileType = 0xFF
+)
+
+const (
+	// FileHeaderMinLength is the minimum length of a firmware file header.
+	FileHeaderMinLength = 0x18
+)
+
+// FirmwareFileHeader represents an EFI File header.
+type FirmwareFileHeader struct {
+	Name           [16]uint8 // This is the GUID of the file.
+	IntegrityCheck [2]uint8
+	Type           FVFileType
+	Attributes     uint8
+	Size           [3]uint8
+	State          uint8
+}
+
+// FirmwareFileHeaderExtended represents an EFI File header with the
+// large file attribute set.
+// We also use this as the generic header for all EFI files, regardless of whether
+// they are actually large. This makes it easier for us to just return one type
+// All sizes are also copied into the ExtendedSize field so we only have to check once
+type FirmwareFileHeaderExtended struct {
+	FirmwareFileHeader
+	ExtendedSize uint64
+}
+
+// FirmwareFile represents an EFI File.
+type FirmwareFile struct {
+	Header FirmwareFileHeaderExtended
+	buf    []byte
+}
+
+// NewFirmwareFile parses a sequence of bytes and returns a FirmwareFile
+// object, if a valid one is passed, or an error. If no error is returned and the FirmwareFile
+// pointer is nil, it means we've reached the volume free space at the end of the FV.
+func NewFirmwareFile(buf []byte) (*FirmwareFile, error) {
```

Comment:
I'll rather stick with the slice now since everything uses the slice. if we want to change it at some point later we can.

I also use the slice directly in a bunch of places.